Fetch Koha shelving locations for every Koha item

alexklbuckley opened the pull request:
Sponsored-By: Auckland University of Technology, New Zealand

demiankatz commented:
@alexklbuckley, what is the difference between location and shelvinglocation? Is this designed to represent permanent vs. temporary location, or something else? Wondering if there's a specific reason to add a new separate value instead of overriding the existing location. If there really is a need for a new value, it might be better to figure out how to incorporate this into the existing ILS driver spec and general code instead of just adding a non-standard value in the KohaRest driver alone.

cmurdoch commented:
@demiankatz The way we use location and shelving location - location is a branch/physical library; shelving location is a permanent location of an item. So we have three branches here with various shelving locations in each.
For practical purposes the shelving location is used in VuFind to display where the item actually is - main collection, high demand collection etc. The only real practical use for the location is as a destination for items in transit. So in VuFind the locations are shown as possible pickup points for requests.

Hope this helps a little. For clarity our usage isn't to do with permanent or temporary location of an item.

demiankatz commented:
Thanks, @cmurdoch. I wonder if what we actually need here is a setting in KohaRest.ini that controls how the location field is populated in the holdings/status results. It seems there are two different types of values that could be used in these contexts, and people are either going to want one or the other. Making the behavior configurable seems like a less complicated solution than exposing both values through separate fields. And changing how locations are displayed in availability/holdings screens doesn't need to have any impact on the existing behavior that uses locations for hold pickup locations, etc., so I think it's a low-risk change as well.

Would you agree? @EreMaijala, what do you think?

alexklbuckley commented:
Hi @EreMaijala , just checking if you've had a chance to take a look at @demiankatz message https://github.com/vufind-org/vufind/pull/3435#issuecomment-2110568493 and what your thoughts are?

Thanks so much,
Alex

EreMaijala commented:
Sorry, I lost track of this. 😞 

Demian's proposal to make the location to display configurable would be my preferred choice as well. Otherwise the location display can get crowded and confusing. But regardless of that I think the code should fetch all locations and cache the values to avoid potentially a terrible number of extra API calls. See `KohaRest::getLibraryName()` for how it's done with libraries.

wainuicatalyst committed:
Changed get location to get the shelving location as well as the branch
location

demiankatz reviewed (changes requested):
If you run `composer fix`, that should correct the style issues that are causing the build to fail.

wainuicatalyst committed:
Merge branch 'vufind-org:dev' into shelvinglocation

wainuicatalyst committed:
Merge branch 'vufind-org:dev' into shelvinglocation

alexklbuckley committed:
Fixing failing unit tests for shelving location

- Indentation fixes
- Declaring variables

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,21 +2470,32 @@ protected function mapRenewalBlockReason($reason, $itype)
-            $libraries = $this->getLibraries();
-            $name = isset($libraries[$libraryId])
+        $name = '';
+        if ($this->config['ItemTypeRenewalBlockMappings']['Location'] == 'Branch') {
```
It would be a good idea to add a default value in case the configuration is undefined -- e.g.

```
if ($this->config['ItemTypeRenewalBlockMappings']['Location'] ?? 'Branch' == 'Branch') {
```

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,21 +2470,32 @@ protected function mapRenewalBlockReason($reason, $itype)
+            $name = $this->translateLocation($libraryId);
+            if ($name === $libraryId) {
+                $libraries = $this->getLibraries();
+                $name = isset($libraries[$libraryId])
```
The line immediately following this one should either be combined with this line or idented one more level beneath it -- right now the ternary operator is not indented, which is not our usual style (though surprisingly, none of our tools seem to mind!).

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,21 +2470,32 @@ protected function mapRenewalBlockReason($reason, $itype)
+                $name = isset($libraries[$libraryId])
                 ? $libraries[$libraryId]['name'] : $libraryId;
+            }
+        } elseif ($this->config['ItemTypeRenewalBlockMappings']['Location'] == 'Shelving') {
```
Maybe the location setting should be set to a variable before this if/else so that you don't have to repeat the default value checking here. Or maybe the whole structure should be a switch to make it easier to add more cases in future.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,21 +2470,32 @@ protected function mapRenewalBlockReason($reason, $itype)
+            if ($name === $shelvingLocationId) {
+                $shelvingLocations = $this->getShelvingLocations();
+                $name = isset($shelvingLocations[$shelvingLocationId])
+                ? $shelvingLocations[$shelvingLocationId]['description'] : $shelvingLocationId;
```
Another unindented ternary.

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley -- this looks pretty good to me apart from a few nitpicks below. As usual, I'll wait for @EreMaijala's input before merging in case he has further comments on the Koha-specific details.

wainuicatalyst committed:
Merge branch 'vufind-org:dev' into shelvinglocation

wainuicatalyst committed:
3435 Switch case for location type

wainuicatalyst committed:
3435 fixing indentation

wainuicatalyst committed:
3435 syntax updates

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,22 +2470,37 @@ protected function mapRenewalBlockReason($reason, $itype)
-            $libraries = $this->getLibraries();
-            $name = isset($libraries[$libraryId])
-                ? $libraries[$libraryId]['name'] : $libraryId;
+        $name = '';
```
Now that your switch is set up in a way that every branch is guaranteed to set `$name` to a value, you don't need to initialize it here.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,22 +2470,37 @@ protected function mapRenewalBlockReason($reason, $itype)
-                ? $libraries[$libraryId]['name'] : $libraryId;
+        $name = '';
+
+        switch ($this->config['ItemTypeRenewalBlockMappings']['Location']) {
```
I still think you should create a default value for the configuration in case it is omitted from the config file. It might also be worth doing some simple normalization to be more tolerant of slight variations. For example, you could do something like:

```
$locationConfig = strtolower(trim($this->config['ItemTypeRenewalBlockMappings']['Location'] ?? 'branch'));
switch ($locationConfig) {
    case 'shelving':
```

That prevents whitespace and case sensitivity problems in addition to providing the legacy behavior as a default. What do you think?

EreMaijala replied:
Maybe it would be better to normalize the config setting in the init method and store it in a property? It might cause it to be done once even if not needed, but would avoid repeating the same code for each item.

demiankatz replied:
That makes sense to me. Adding a new property that gets initialized in the init() method might improve readability elsewhere. What do you think, @alexklbuckley?

alexklbuckley replied:
I agree! I've added that in my latest follow-up. thanks both!

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,22 +2470,37 @@ protected function mapRenewalBlockReason($reason, $itype)
+                $name = $this->translateLocation($shelvingLocationId);
+                if ($name === $shelvingLocationId) {
+                    $shelvingLocations = $this->getShelvingLocations();
+                    $name = isset($shelvingLocations[$shelvingLocationId])
```
I believe you could simplify this to:

```
$name = $shelvingLocations[$shelvingLocationId]['description'] ?? $shelvingLocationId;
```

That also protects against the (probably unlikely, but nothing is impossible) situation where the data is set by the ID but somehow missing the location element.

demiankatz replied:
(And I see now that this code is based on pre-existing code for processing library names below... but I still think it might be worth improving, and you could make a matching change to the name logic).

alexklbuckley replied:
thanks @demiankatz . Sorry I missed this change earlier. I've gone ahead and fixed this in my follow-up commit.

demiankatz reviewed (changes requested):
Looking much better, @alexklbuckley! A few more nitpicks (some optional) in case you want to further polish while we wait for @EreMaijala to get back and finish reviewing. :-)

wainuicatalyst committed:
Merge branch 'vufind-org:dev' into shelvinglocation

wainuicatalyst committed:
3425 additional changes

alexklbuckley committed:
Shelving location display: Fixes for failing unit tests

- Fix for instantiating a variable
- Indentation fixes

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2449,22 +2470,36 @@ protected function mapRenewalBlockReason($reason, $itype)
-                ? $libraries[$libraryId]['name'] : $libraryId;
+        $locationConfig = strtolower(trim($this->config['ItemTypeRenewalBlockMappings']['Location'] ?? 'branch'));
+        switch ($locationConfig) {
+            case 'Shelving':
```
If you `strtolower` the value, you need to change this from `Shelving` to `shelving` or it will never match. (That change was included in my earlier example, but I should have highlighted it since it's not very obvious at a glance -- sorry about that).

alexklbuckley replied:
thanks very much @demiankatz . Sorry about the delay, I have pushed a fix changing this case back to lower case.

demiankatz reviewed (changes requested):
A small problem was introduced by the latest edits -- see below. I think there's also one outstanding comment from the past review which I have left unresolved.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into shelvinglocation

alexklbuckley committed:
Shelving location display: Change case statement to lower case

demiankatz commented:
Thanks, @alexklbuckley! I think this can probably be approved once @EreMaijala returns, but I'll wait for his input before proceeding. I do see that I still have an open improvement suggestion above, which is definitely not required, but might be helpful for readability.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into shelvinglocation

alexklbuckley committed:
Shelving location display: Simplify how the library and shelving
location name is set

alexklbuckley commented:
Thanks @demiankatz I've just pushed a fix for that open improvement suggestion above, so that should be everything covered now. I appreciate your help!

demiankatz commented:
Thanks, @alexklbuckley, I agree -- looks like we're all caught up and ready for @EreMaijala's feedback when he gets back (and caught up). :-)

alexklbuckley commented:
Thanks @demiankatz !

EreMaijala commented on config/vufind/KohaRest.ini:
```diff
@@ -184,3 +184,9 @@ extraFields = item-issue:acceptTerms:pickUpLocation
+; This section controls how the location field in the holdings/status results is populated.
+; It can either be Branch (Koha library branch/physical location) or Shelving (Koha
+; permanent shelving location of an item)
+[LocationFieldForHoldingStatus]
```
Looking at this with fresh eyes raises a question: Why isn't this just a single setting in the Holdings section?

alexklbuckley replied:
thanks @EreMaijala , I've added this in my latest follow-up

alexklbuckley committed:
Merge branch 'vufind-org:dev' into shelvinglocation

alexklbuckley committed:
Shelving location display: Implement review feedback

- Set locationConfig in KohaRest.php init() and store in property
- Move Location to be a single setting in [Holdings] in KohaRest.ini

Sponsored-By: Auckland University of Technology, New Zealand

alexklbuckley commented:
thanks for your feedback @EreMaijala . I hope I have addressed both of your points in my latest follow-up commit. Please let me know :)

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -327,6 +338,9 @@ public function init()
             = $this->config['Holdings']['sortBySerialIssue'] ?? true;
 
+        $this->locationConfig
+        = strtolower(trim($this->config['ItemTypeRenewalBlockMappings']['Location'] ?? 'branch'));
```
This line should be indented, and the section name isn't quite right. :)

Please make the setting start with lower-case letter. And perhaps make it `locationField` to be a bit more specific?

I just realized that the example KohaRest.ini doesn't actually contain the Holdings section (oops!). Sorry about the confusion this may have caused. I'll put it on my TODO list to augment it later.

alexklbuckley replied:
thanks @EreMaijala ! I've added those changes in my latest follow-up. 

I'm not sure if my addition of [Holdings] to KohaRest.ini is enough?

alexklbuckley committed:
Merge branch 'vufind-org:dev' into shelvinglocation

alexklbuckley committed:
Shelving location display: Implement review feedback

- Change name, indentation and case of KohaRest location field config

Sponsored-By: Auckland University of Technology, New Zealand

demiankatz reviewed (approved):
Thanks, @alexklbuckley and @EreMaijala!

alexklbuckley commented:
Lovely, thanks @demiankatz and @EreMaijala !
